src/st/no0/e_slinger.c

```diff
@@ -0,0 +1,231 @@
+static u16 dead_parts_pos_y[] = {-16, -8, -4, -4, 9, 9, 0, 0};
+static u8 attack_timer_cycles[2][4] = {
+    {0x80, 0x08, 0x08, 0x40}, {0xF0, 0xC0, 0xA0, 0x80}};
+static s16 sensors_ground[][2] = {{0, 19}, {8, 0}};
+static s16 sensors_move[][2] = {{-12, 16}, {0, -16}, {0, -16}};
+
+void SlingerAttackCheck(Entity* self) {
+    s32 groundCollision = UnkCollisionFunc2(sensors_ground);
+    s16 moveCollision = UnkCollisionFunc(sensors_move, 3);
+
+    if ((groundCollision & 0xFFFF) == 128 && (moveCollision & 2) != 0) {
+        SetStep(SLINGER_JUMP);
+    } else if (--self->ext.skeleton.attackTimer == 0) {
+        SetStep(SLINGER_ATTACK);
+    }
+}
+
+extern u16 g_EInitSlinger[];
+void EntitySlinger(Entity* self) {
+    Entity* newEntity;
+    u8 animStatus;
+    u8 i;
+
+    if (self->flags & FLAG_DEAD) {
+        self->step = SLINGER_DESTROY;
+    }
+
+    switch (self->step) {
+    case SLINGER_INIT:
+        InitializeEntity(g_EInitSlinger);
+        self->ext.skeleton.attackTimer = 80; // Slinger attack timer cycle
+        self->ext.skeleton.facingLeft = 0;   // Facing init
+        self->ext.skeleton.attackTimerIndex = 0;
+        break;
+    case SLINGER_IDLE: // Wait for player to be close enough
+        if (UnkCollisionFunc3(sensors_ground) != 0) {
+            self->step++;
+        }
+        break;
+    case SLINGER_WALK_TOWARDS_PLAYER:
+        self->facingLeft = (GetSideToPlayer() & 1) ^ 1;
+        self->ext.skeleton.facingLeft = self->facingLeft;
+        AnimateEntity(anim_walk, self);
+
+        if (self->ext.skeleton.facingLeft == 0) {
+            self->velocityX = FIX(-0.5);
+        } else {
+            self->velocityX = FIX(0.5);
+        }
+
+        if (GetDistanceToPlayerX() < 76) {
+            self->step = SLINGER_WALK_AWAY_FROM_PLAYER;
+        }
+        SlingerAttackCheck(self);
+        break;
+    case SLINGER_WALK_AWAY_FROM_PLAYER:
+        self->facingLeft = (GetSideToPlayer() & 1) ^ 1;
+        self->ext.skeleton.facingLeft = self->facingLeft ^ 1;
+        AnimateEntity(anim_walk_backwards, self);
+
+        if (self->ext.skeleton.facingLeft == 0) {
+            self->velocityX = FIX(-0.5);
+        } else {
+            self->velocityX = FIX(0.5);
+        }
+
+        if (GetDistanceToPlayerX() > 92) {
+            self->step = SLINGER_WALK_TOWARDS_PLAYER;
+        }
+        SlingerAttackCheck(self);
+        break;
+    case SLINGER_ATTACK:
+        animStatus = AnimateEntity(anim_throw_bone, self);
+        if (!animStatus) {
+            SetStep(SLINGER_WALK_AWAY_FROM_PLAYER);
+            self->ext.skeleton.attackTimerIndex++;
+            self->ext.skeleton.attackTimer =
+                attack_timer_cycles[self->params & 1]
+                                   [self->ext.skeleton.attackTimerIndex & 3];
+            break;
+        }
+
+        if ((animStatus & 0x80) && (self->animCurFrame == 10)) {
+            if (((u32)(((u16)self->posX.i.hi + 16) & 0xFFFF) <= 288) &&
```

Comment:
This is
```
            if ((self->posX.i.hi >= -16 && self->posX.i.hi <= 272) &&
                (self->posY.i.hi >= 0 && self->posY.i.hi <= 240)) {
```
https://decomp.me/scratch/N3g0Z

Reply:
Thanks, I'll fix this up. Just starting to see a bit of this min <= x <= max pattern in this code for the first time I think